### PR TITLE
[python-package] Add tests on Booster.eval() handling of new validation sets

### PR DESCRIPTION
**Pull request by daguirre11**
Contributes to https://github.com/microsoft/LightGBM/issues/7031
**BEFORE**
Screenshot 2026-02-24 at 1 31 13 PM

**AFTER**
Screenshot 2026-02-24 at 1 23 14 PM

2 line difference in coverage for `/lightgbm/basic.py`

### CODE REVIEW

File: tests/python_package_test/test_basic.py [hunk 1]
```diff
@@ -1098,3 +1098,23 @@ def test_set_field_none_removes_field(rng, field_name):
...
+        *load_breast_cancer(return_X_y=True),
+        test_size=0.1,
+        random_state=42,
+    )
+    train_set = lgb.Dataset(X_train, label=y_train)
+    valid_set = lgb.Dataset(X_test, label=y_test, reference=train_set)
+    booster = lgb.Booster(
+        params={"objective": "binary", "verbose": -1},
+        train_set=train_set,
+    )
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py:1124]:
```suggestion
    booster = lgb.Booster(
        params={
            "deterministic": True,
            "force_row_wise": True,
            "objective": "binary",
            "metric": ["auc", "binary_error"],
            "num_iterations": 2,
            "num_leaves": 3,
            "num_threads": 1,
            "seed": 708,
            "verbose": -1,
        },
        train_set=train_set,
    )
```

Some suggested changes:

* make the model tiny (and therefore fast)
* set multiple values for `metric` to check that `.eval()` uses all of them
* make the model deterministic

File: tests/python_package_test/test_basic.py [hunk 2]
```diff
@@ -1098,3 +1098,23 @@ def test_set_field_none_removes_field(rng, field_name):
...
+    num_datasets_before = booster._Booster__num_dataset
+    result = booster.eval(valid_set, name="test")
+
+    assert booster._Booster__num_dataset == num_datasets_before + 1
+    assert len(result) > 0
+    assert result[0][0] == "test"
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py]:
```suggestion
    assert booster.__num_dataset == 1
    assert booster.valid_sets == []

    result = booster.eval(valid_set, name="test")

    assert booster._Booster__num_dataset == 2
    assert booster.valid_sets == [valid_set]
    assert len(result) == 2
    assert isinstance(result, list)

    # first metric - AUC
    dataset_name, metric_name, metric_value, maximize = result[0]
    assert dataset_name == "test"
    assert metric_name == "auc"
    assert metric_value >= 0.50
    assert maximize = True

    # second metric - binary error
    dataset_name, metric_name, metric_value, maximize = result[1]
    assert dataset_name == "test"
    assert metric_name == "binary_error"
    assert metric_value >= 0.40
    assert maximize = True
```

The test you set up is great for covering the main thing we want to cover (that `Booster.eval()`is adding that Dataset it's passed to `.valid_sets`), so please don't take this as a criticism. I just think right here, now that we have this test, we have an opportunity to make it much stricter without too much extra complexity.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks so much for doing this, it's great to have some more test coverage of this method!

File: tests/python_package_test/test_basic.py [hunk 3]
```diff
@@ -1098,3 +1098,50 @@ def test_set_field_none_removes_field(rng, field_name):
...
+    assert metric_name == "auc"
+    assert metric_value >= 0.50
+    assert maximize = True
+
+    # second metric - binary error
+    dataset_name, metric_name, metric_value, maximize = result[1]
+    assert dataset_name == "test"
+    assert metric_name == "binary_error"
+    assert metric_value >= 0.40
+    assert maximize = True
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py]:
```suggestion
    assert maximize is True

    # second metric - binary error
    dataset_name, metric_name, metric_value, maximize = result[1]
    assert dataset_name == "test"
    assert metric_name == "binary_error"
    assert metric_value >= 0.40
    assert maximize is True
```

File: tests/python_package_test/test_basic.py [hunk 4]
```diff
@@ -1098,3 +1098,50 @@ def test_set_field_none_removes_field(rng, field_name):
...
+            "metric": ["auc", "binary_error"],
+            "num_iterations": 2,
+            "num_leaves": 3,
+            "num_threads": 1,
+            "seed": 708,
+            "verbose": -1,
+        },
+        train_set=train_set,
+    )
+    assert booster.__num_dataset == 1
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py]:
```suggestion
    assert booster._Booster__num_dataset == 1
```

File: tests/python_package_test/test_basic.py [hunk 5]
```diff
@@ -1098,3 +1098,50 @@ def test_set_field_none_removes_field(rng, field_name):
...
+    assert metric_name == "auc"
+    assert metric_value >= 0.50
+    assert maximize is True
+
+    # second metric - binary error
+    dataset_name, metric_name, metric_value, maximize = result[1]
+    assert dataset_name == "test"
+    assert metric_name == "binary_error"
+    assert metric_value >= 0.40
+    assert maximize is True
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py]:
```suggestion
    assert maximize is False
```

**Comment by jameslamb** [comment]
Sorry for the couple of bad commits I pushed here, I think everything's looking good now!

The only remaining failing CI is the job we use to test against the latest nightly packages or `lightgbm`'s dependencies. That's failing for reasons unrelated to this PR, tracked here: https://github.com/scikit-learn/scikit-learn/issues/33616

If that isn't resolved upstream in a few days, we'll temporarily turn off that job so PRs like this one can move forwad.
